iml-api/src/graphql.rs

```diff
@@ -478,6 +559,176 @@ impl QueryRoot {
 
         Ok(xs)
     }
+
+    #[graphql(arguments(
+        limit(description = "optional paging limit, defaults to all rows"),
+        offset(description = "Offset into items, defaults to 0"),
+        dir(description = "Sort direction, defaults to asc"),
+        message(description = "Pattern to search for in message"),
+        fqdn(description = "Pattern to search for in FQDN"),
+        tag(description = "Pattern to search for in tag"),
+        start_datetime(description = "Start of the time period of logs"),
+        end_datetime(description = "End of the time period of logs"),
+        message_class(description = "Array of log message classes"),
+        severity(description = "Upper bound of log severity"),
+    ))]
+    async fn logs(
+        context: &Context,
+        limit: Option<i32>,
+        offset: Option<i32>,
+        dir: Option<SortDir>,
+        message: Option<String>,
+        fqdn: Option<String>,
+        tag: Option<String>,
+        start_datetime: Option<String>,
+        end_datetime: Option<String>,
+        message_class: Vec<MessageClass>,
+        severity: Option<LogSeverity>,
+    ) -> juniper::FieldResult<Vec<LogMessage>> {
+        let dir = dir.unwrap_or_default();
+
+        let start_datetime: Option<chrono::DateTime<Utc>> = if let Some(s) = start_datetime {
+            Some(s.parse()?)
+        } else {
+            None
+        };
```

Comment:
something like:
```suggestion
        let start_datetime: Option<chrono::DateTime<Utc>> = start_datetime.map(|s| s.parse()?);
```
?

Reply:
Nah, map creates a closure and `?` stops working.

Reply:
https://doc.rust-lang.org/stable/rust-by-example/error/multiple_error_types/option_result.html ?

Reply:
Yes, but it coerces all parsing errors to omission of argument:
```
        let start_datetime = start_datetime
            .map(|s| s.parse::<chrono::DateTime<Utc>>())
            .map_or(Ok(None), |r| r.map(Some))?;
```

Currently it's like this:
![image](https://user-images.githubusercontent.com/393741/95185199-15c93800-07d1-11eb-8f69-fc2d48db5bae.png)

Reply:
Hm, maybe I'm wrong, let me double-check

Reply:
Ok, it actually works. Thanks for suggestion!